File: sentimental-credit/credit.py
def main():

    card_number = input("Enter credit card number: ")

    # Substring in python source: https://www.freecodecamp.org/news/how-to-substring-a-string-in-python/

    first_set_of_numbers = card_number[-2::-2]
    sum = 0

    # Multiply by 2 every number in the first set
    for number in first_set_of_numbers:

        number = int(number) * 2

        # If number is greater than 9 (means it is a 2 digit number) sum every digit individually
        # else add number to the sum
        if number > 9:
            sum += int(str(number)[0:1])
            sum += int(str(number)[1:2])
        else:

            sum += number

    second_set_of_numbers = card_number[-1::-2]

    # Sum every number in the second set
    for number in second_set_of_numbers:

        sum += int(number)

    # If last digit of sum is 0 start checking for card's circuit accordingly else print INVALID
    if str(sum)[-1] == str(0):

        first_digit = int(card_number[0:1])
        first_two_digits = int(card_number[0:2])
        card_number_length = len(card_number)

        if card_number_length == 15 and first_two_digits == 34 or first_two_digits == 37:
            print("AMEX")

        elif card_number_length == 16 and first_two_digits >= 51 and first_two_digits <= 55:
            print("MASTERCARD")

        elif card_number_length == 13 or card_number_length == 16 and first_digit == 4:
            print("VISA")

        else:
            print("INVALID")

    else:
        print("INVALID")


main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int getLast2Digits(long cardNumber);
int getLastDigit(long cardNumber);

int main(void)
{
    // Initialize variables
    long cardNumber = get_long("Enter credit card number: ");

    // Get a copy of cardNumber so the algorithm can use this later to check for digits
    long cardNumberCopy = cardNumber;
    int lastDigit = 0;
    int evenDigitsSum = 0;
    int splittedDigit = 0;
    int evenDigitMultiplied = 0;
    int oddDigitsSum = 0;
    int last2Digits = 0;
    int i;

    for (i = 1; cardNumberCopy + 1 / 10 > 0; i++)
    {
        lastDigit = cardNumberCopy % 10;

        cardNumberCopy /= 10;

        // Multipy digits starting from the second one by 2
        if (i % 2 == 0)
        {
            evenDigitMultiplied = lastDigit * 2;

            // If multiplied digit results in a double digit, split and add into evenDigitsSum variable
            // example: 7 * 2 = 14, split the 14 into 1 + 4 = 5 and sum it into evenDigitsSum variable
            while (evenDigitMultiplied >= 10)
            {
                splittedDigit = evenDigitMultiplied % 10;
                evenDigitMultiplied /= 10;
                evenDigitsSum += splittedDigit;
            }

            evenDigitsSum += evenDigitMultiplied;
        }
        // Get the remaining digits and sum it into oddDigitsSum variable
        else
        {
            oddDigitsSum += lastDigit;
        }
    }

    // Decrement counter by i since it has been intialized as 1 in the previous for loop
    i--;

    // Check if the last digit of the sum of even and odd (position) numbers is 0
    if ((evenDigitsSum + oddDigitsSum) % 10 == 0)
    {
        if (i == 15)
        {
            last2Digits = getLast2Digits(cardNumber);

            if (last2Digits == 34 || last2Digits == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (i == 16)
        {
            // Get last 2 and last digit
            last2Digits = getLast2Digits(cardNumber);
            lastDigit = getLastDigit(cardNumber);

            // If last 2 digits are between 51 and 55 then it's a mastercard
            if (last2Digits >= 51 && last2Digits <= 55)
            {
                printf("MASTERCARD\n");
            }
            // If last digit is 4 then it's a visa
            else if (lastDigit == 4)
            {
                printf("VISA\n");
            }
            // else card is not valid
            else
            {
                printf("INVALID\n");
            }
        }
        else if (i == 13 && getLastDigit(cardNumber) == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

// Source code for finding first 2 digits of a number: https://stackoverflow.com/questions/52489462/how-to-find-first-two-digits-in-c
int getLast2Digits(long cardNumber)
{
    while (cardNumber >= 100)
    {
        cardNumber /= 10;
    }

    return cardNumber;
}

// Source code for finding first digit of a number: https://codeforwin.org/2015/06/how-to-find-first-and-last-digit-of-any-number.html
int getLastDigit(long cardNumber)
{
    while (cardNumber >= 10)
    {
        cardNumber /= 10;
    }

    return cardNumber;
}
